### PR TITLE
FieldMapper

### LINKED ISSUE #38: Implements a FieldMapper for the statement generator
We need a `FieldMapper` that maps a `case class` into a list with the field names and the values exported to ByteBuffer values.

### PR DESCRIPTION
**Pull request by fedefernandez**
Fixes #38

### CODE REVIEW

File: core/src/main/scala/mapper/ByteBufferMapper.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle.cassandra
+package mapper
+
+trait ByteBufferMapper[A] {
+  def map(a: A): List[(MappedField)]
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/ByteBufferMapper.scala]: Why do not just `List[MappedField]`?

File: core/src/main/scala/mapper/ByteBufferMapper.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+object ByteBufferMapper {
+
+  def apply[A](implicit sg: ByteBufferMapper[A]): ByteBufferMapper[A] =
+    sg
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/ByteBufferMapper.scala]: `sg` -> `ev`?

File: core/src/main/scala/mapper/ByteBufferMapper.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  implicit def genericMapper[A](implicit fieldLister: FieldMapper[A]): ByteBufferMapper[A] =
+    new ByteBufferMapper[A] {
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/ByteBufferMapper.scala]: Could we use just the abstract method?

```scala
implicit def genericMapper[A](implicit FM: FieldMapper[A]): ByteBufferMapper[A] =
    (a: A) => FM.map(a)
```

`fieldLister` -> `FM`

File: core/src/main/scala/mapper/package.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import freestyle.cassandra.codecs.ByteBufferCodec
+import shapeless._
+import shapeless.labelled.FieldType
+
+package object mapper {
+
+  type MappedField = (String, ByteBuffer)
+
+  trait FieldMapper[A] {
+    def map(a: A): List[((MappedField))]
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/package.scala]: `List[MappedField]`?

File: core/src/main/scala/mapper/package.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  trait FieldMapperPrimitive {
+
+    def createFieldMapper[A](f: A => List[((MappedField))]): FieldMapper[A] =
+      new FieldMapper[A] {
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/package.scala]: Same as above. Also, what about using this?

```scala
def createFieldMapper[A](f: A => List[MappedField]): FieldMapper[A] = (a: A) => f(a)
```

File: core/src/main/scala/mapper/package.scala [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+  trait FieldMapperGeneric extends FieldMapperPrimitive {
+
+    implicit def genericMapper[A, R](
+        implicit gen: LabelledGeneric.Aux[A, R],
+        mapper: Lazy[FieldMapper[R]]): FieldMapper[A] =
+      createFieldMapper(value => mapper.value.map(gen.to(value)))
+
+    implicit val hnilLister: FieldMapper[HNil] = new FieldMapper[HNil] {
+      override def map(a: HNil): List[((MappedField))] = Nil
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/package.scala]: `List[MappedField]`

File: core/src/main/scala/mapper/package.scala [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  trait FieldMapperGeneric extends FieldMapperPrimitive {
+
+    implicit def genericMapper[A, R](
+        implicit gen: LabelledGeneric.Aux[A, R],
+        mapper: Lazy[FieldMapper[R]]): FieldMapper[A] =
+      createFieldMapper(value => mapper.value.map(gen.to(value)))
+
+    implicit val hnilLister: FieldMapper[HNil] = new FieldMapper[HNil] {
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/package.scala]: `hnilMapper`?

Also, same as mentioned earlier:
```scala
implicit val hnilMapper: FieldMapper[HNil] = (a: HNil) => Nil
```

File: core/src/main/scala/mapper/package.scala [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    def createFieldMapper[A](f: A => List[((MappedField))]): FieldMapper[A] =
+      new FieldMapper[A] {
+        override def map(a: A): List[((MappedField))] = f(a)
+      }
+
+    implicit def primitiveFieldMapper[K <: Symbol, H, T <: HList](
+        implicit witness: Witness.Aux[K],
+        codec: Lazy[ByteBufferCodec[H]],
+        tLister: FieldMapper[T]): FieldMapper[FieldType[K, H] :: T] = {
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/package.scala]: `tMapper`?

File: core/src/main/scala/mapper/package.scala [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      override def map(a: HNil): List[((MappedField))] = Nil
+    }
+
+  }
+
+  object FieldMapper extends FieldMapperGeneric
+
+  object FieldMapperExpanded extends FieldMapperGeneric {
+
+    implicit def hconsLister[K, H, T <: HList](
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/package.scala]: `hconsMapper`?

File: core/src/main/scala/mapper/package.scala [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        implicit hLister: Lazy[FieldMapper[H]],
```
Comment (juanpedromoreno) [inline core/src/main/scala/mapper/package.scala]: `hMapper`

File: core/src/test/scala/mapper/ByteBufferMapperSpec.scala [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+  case class C(c1: String, c2: B)
+
+  implicit val stringTypeCodec: TypeCodec[String] = TypeCodec.varchar()
+  implicit val protocol: ProtocolVersion          = ProtocolVersion.V3
+  import codecs._
+
+  val stringCodec: ByteBufferCodec[String] = implicitly[ByteBufferCodec[String]]
+
+  "map" should {
```
Comment (juanpedromoreno) [inline core/src/test/scala/mapper/ByteBufferMapperSpec.scala]: `map` is too generic, what about naming it as `FieldMapper.map`?

File: core/src/main/scala/mapper/package.scala (hunk 5 above)
Comment (fedefernandez) [reply core/src/main/scala/mapper/package.scala]: it only works on Scala 2.12 :(

**Review by juanpedromoreno — APPROVED** [review]
LGTM, good job @fedefernandez ! Thanks.

giphy 16

**Pull request merged**